kratos/containers/constraint_data.h

```diff
@@ -0,0 +1,407 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:		 BSD License
+//					 Kratos default license: kratos/license.txt
+//
+//  Main authors:    Aditya Ghantasala
+//
+//
+
+#if !defined(CONSTRAINT_DATA_H)
+#define CONSTRAINT_DATA_H
+// System includes
+#include <vector>
+#include <unordered_map>
+#include <iostream>
+#include <tuple>
+#include <utility>
+#include <assert.h>
+
+// project includes
+#include <boost/functional/hash.hpp>
+#include <boost/multi_index_container.hpp>
+#include <boost/multi_index/composite_key.hpp>
+#include <boost/multi_index/hashed_index.hpp>
+#include <boost/multi_index/ordered_index.hpp>
+#include <boost/multi_index/member.hpp>
+#include <boost/range/iterator_range.hpp>
+#include "includes/define.h"
+#include "includes/dof.h"
+#include "includes/node.h"
+
+namespace Kratos
+{
+
+// ConstraintEquation class start
+/*
+*   This class stores the information regarding the constraint equation. Naming convenction is defined like this.
+*   (each object of this class will store one equation in the given form)
+*
+*   slaveDOF = w_1*masterDOF_1 + w_2*masterDOF_2 + ..... + w_n*masterDOF_n
+*   
+*   each slaveDOF and each of its masterDOF have the following attributes 
+*   a. dof ID
+*   b. dof KEY
+*   c. equation ID
+*
+*   one should be able to access the constraint equation both with (dofID , dofKEY) pair and/or equationID of the slave.
+*   This equation is imposed on the linear system of equations either element wise or on the global system.
+*
+*/
+class ConstraintEquation
+{
+  private:
+    struct MasterData
+    {
+
+        MasterData()
+        {
+            masterDofId = -1;
+            masterKey = -1;
+            masterWeight = 0.0;
+            masterEqId = -1;
+        }
+        MasterData(unsigned int iMasterDofId, unsigned int iMasterKey, double iWeight)
+        {
+            masterDofId = iMasterDofId;
+            masterKey = iMasterKey;
+            masterWeight = iWeight;
+            masterEqId = -1;
+        }
+        MasterData(unsigned int iequationId, double iWeight)
+        {
+            masterDofId = -1;
+            masterKey = -1;
+            masterEqId = iequationId;
+            masterWeight = iWeight;
+        }
+        unsigned int masterDofId;
+        unsigned int masterKey;
+        double masterWeight;
+        unsigned int masterEqId;
+    };
+
+    std::vector<MasterData> mMasterData;
+
+    unsigned int slaveDofId;

```

Comment:
Both master and slave have related data in common: DofId, DofKey and EquationId. What about making this a struct called something like ConstraintData and defining a comparison operator for it == in addition to the default assignment operator =. Then use it for both the slave and master. You can add the master weight in a subclass or as a tuple (ConstraintData, double). This would also simplify your search algorithms/code below since you would do equality comparison with the new == operator.